src/ucs/sys/string.c

```diff
@@ -359,26 +359,31 @@ size_t ucs_string_common_prefix_len(const char *str1, const char *str2)
     return (p1 - str1);
 }
 
-ssize_t ucs_path_calc_distance(const char *path1, const char *path2)
+void ucs_path_get_common_parent(const char *path1, const char *path2,
+                                char *common_path, size_t max_length)
 {
-    unsigned distance = 0;
-    size_t common_length;
-    char rpath1[PATH_MAX], rpath2[PATH_MAX];
+    size_t offset, end_offset;
 
-    if ((NULL == realpath(path1, rpath1)) ||
-        (NULL == realpath(path2, rpath2))) {
-        return UCS_ERR_INVALID_PARAM;
+    end_offset = 0;
+    for (offset = 0; path1[offset] == path2[offset]; ++offset) {
+        if (path1[offset] == '/') {
+            end_offset = offset;
+        }
     }
 
-    common_length = ucs_string_common_prefix_len(rpath1, rpath2);
-
-    if (rpath1[common_length] != rpath2[common_length]) {
-        ++distance; /* count the differentiating path component */
-    }
+    end_offset = ucs_min(end_offset, max_length - 1);
+    memcpy(common_path, path1, end_offset);
+    common_path[end_offset] = '\0';
+}
 
-    distance += ucs_max(ucs_string_count_char(rpath1 + common_length, '/'),
-                        ucs_string_count_char(rpath2 + common_length, '/'));
+size_t ucs_path_calc_distance(const char *path1, const char *path2)
+{
+    size_t common_length = ucs_string_common_prefix_len(path1, path2);
+    size_t distance;
 
+    distance  = (path1[common_length] == path2[common_length]) ? 0 : 1;
+    distance += ucs_max(ucs_string_count_char(path1 + common_length, '/'),
+                        ucs_string_count_char(path2 + common_length, '/'));
```

Comment:
Do you I understand correctly that if paths are /a/b/c and /a/x/y then distance is 2 but if paths are /a/b/c and a/p/q/r then distance is 3. Shouldn't the distances be 4 and 5 for the previous two cases respectively? (I'm assuming last c, y, r are the sys devices)

Reply:
it's a mistake - pushing a fix..

Reply:
done